zebra/zebra_ns_id.c

```diff
@@ -0,0 +1,360 @@
+	addattr32(nlh, NETLINK_SOCKET_BUFFER_SIZE, NETNSA_NSID, ns_id);
+
+	ret = send_receive(sock, nlh, seq, buf);
+	if (ret < 0) {
+		close(sock);
+		close(fd);
+		return NS_UNKNOWN;
+	}
+	nlh = (struct nlmsghdr *)buf;
+
+	/* message to analyse : NEWNSID response */
+	len = ret;
+	ret = 0;
+	do {
+		if (nlh->nlmsg_type >= NLMSG_MIN_TYPE) {
+			return_nsid = extract_nsid(nlh, buf);
+			if (return_nsid != NS_UNKNOWN)
+				break;
+		} else {
+			if (nlh->nlmsg_type == NLMSG_ERROR) {
+				struct nlmsgerr *err = (struct nlmsgerr *)
+					((char *)nlh + NETLINK_ALIGN(sizeof(struct nlmsghdr)));
+
+				ret = -1;
+				if (err->error < 0)
+					errno = -err->error;
+				else
+					errno = err->error;
+				if (errno == 0) {
+					/* request NEWNSID was successfull
+					 * return EEXIST error to get GETNSID
+					 */
+					errno = EEXIST;
+				}
+			} else {
+				/* other errors ignored
+				 * attempt to get nsid
+				 */
+				ret = -1;
+				errno = EEXIST;
+				break;
+			}
+		}
+		len = len - NETLINK_ALIGN(nlh->nlmsg_len);
+		nlh = (struct nlmsghdr *)((char *)nlh + NETLINK_ALIGN(nlh->nlmsg_len));
+	} while (len != 0 && return_nsid != NS_UNKNOWN && ret == 0);
+
+	if (ret <= 0) {
+		if (errno != EEXIST && ret != 0) {
+			zlog_err("netlink( %u) recvfrom() error 2 when reading: %s",
+				 fd, safe_strerror(errno));
+			close(sock);
+			close(fd);
+			if (errno == ENOTSUP) {
+				zlog_warn("NEWNSID locally generated");
+				return zebra_ns_id_get_fallback(netnspath);
+			}
+			return NS_UNKNOWN;
+		}
+		/* message to send to netlink : GETNSID */
+		memset(buf, 0, NETLINK_SOCKET_BUFFER_SIZE);
+		nlh = initiate_nlh(buf, &seq, RTM_GETNSID);
+		rt = (struct rtgenmsg *)(buf + nlh->nlmsg_len);
+		nlh->nlmsg_len+= NETLINK_ALIGN(sizeof(struct rtgenmsg));
+		rt->rtgen_family = AF_UNSPEC;
+
+		addattr32(nlh, NETLINK_SOCKET_BUFFER_SIZE, NETNSA_FD, fd);
+		addattr32(nlh, NETLINK_SOCKET_BUFFER_SIZE, NETNSA_NSID, ns_id);
+
+		ret = send_receive(sock, nlh, seq, buf);
+		if (ret < 0) {
+			close(sock);
+			close(fd);
+			return NS_UNKNOWN;
+		}
+		nlh = (struct nlmsghdr *)buf;
+		len = ret;
+		ret = 0;
+		do {
+			if (nlh->nlmsg_type >= NLMSG_MIN_TYPE) {
+				return_nsid = extract_nsid(nlh, buf);
+				if (return_nsid != NS_UNKNOWN)
+					break;
+			} else if (nlh->nlmsg_type == NLMSG_ERROR) {
+				struct nlmsgerr *err = (struct nlmsgerr *)
+					((char *)nlh + NETLINK_ALIGN(sizeof(struct nlmsghdr)));
+				if (err->error < 0)
+					errno = -err->error;
+				else
+					errno = err->error;
+				break;
+			}
+			len = len - NETLINK_ALIGN(nlh->nlmsg_len);
+			nlh = (struct nlmsghdr *)((char *)nlh + NETLINK_ALIGN(nlh->nlmsg_len));
+		} while (len != 0 && return_nsid != NS_UNKNOWN && ret == 0);
+	}
+
+	close(fd);
+	close(sock);
+	return return_nsid;
+}
+
+#else
+ns_id_t zebra_ns_id_get(const char *netnspath)
+{
+	return zebra_ns_id_get_fallback(netnspath);
+}
+#endif /* ! defined(HAVE_NETLINK) */
+
+
+static void zebra_ns_create_netns_folder(void)
+{
+#ifdef HAVE_NETNS
```

Comment:
nit: why not put the entire function inside the `#ifdef`? And the "folder" nomenclature seems a bit odd for me, "dir" or "directory" would feel more appropriate IMO. But maybe I'm being too pedantic here :)

Reply:
ok. i will do it.